### PR TITLE
Using random selection to avoid fake tips

### PR DESCRIPTION
**Pull request by rmoreliovlabs**
## Description
Using random selection in order to avoid fake tips.

## Motivation and Context
The problem here is that during long sync a dishonest peer can lie about its total difficulty and then as a result become the one the node is trying to sync against. Even more, the dishonest peer can repeat this again and again on each synchronisation attempt.

Another problem here might be that even with honest peers, the load from peers that are performing full long sync may be directed to the peer(s) which in most of the times report higher difficulty (bootstrap / miner nodes ? )

One way to tackle this situation is to randomly select a peer for syncing from top X% with highest difficulty available (X should be configurable). Note: preferably this new algorithm should have a feature toggle and be disabled by default (eg. when X == 0).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

File: rskj-core/src/main/java/org/ethereum/config/SystemProperties.java [hunk 1]
```diff
@@ -98,6 +98,7 @@ public abstract class SystemProperties {
 
     public static final String PROPERTY_PERSIST_STATES_CACHE_SNAPSHOT = "cache.states.persist-snapshot";
     public static final String PROPERTY_PERSIST_BLOOMS_CACHE_SNAPSHOT = "cache.blooms.persist-snapshot";
+    public static final String PROPERTY_PERCENTAGE_OF_PEERS_TO_CONSIDER_IN_RND_SELECTION = "rpc.percentageOfPeersToConsiderInRandomSelection";
```
Comment (Vovchyk) [inline rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:101]: why `rpc`? how about `sync.topBest`? Also this config has to be added to reference.conf and expected.conf

File: rskj-core/src/main/java/org/ethereum/config/SystemProperties.java [hunk 2]
```diff
@@ -716,4 +717,12 @@ public long getGasEstimationCap() {
     public DbKind databaseKind() {
         return DbKind.ofName(configFromFiles.getString(KeyValueDataSource.KEYVALUE_DATASOURCE_PROP_NAME));
     }
+
+    public double getPercentageOfPeersToConsiderInRandomSelection() {
+        if (!configFromFiles.hasPath(PROPERTY_PERCENTAGE_OF_PEERS_TO_CONSIDER_IN_RND_SELECTION)) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:722]: debugging stuff? we don't need this, right?

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java [hunk 3]
```diff
@@ -51,11 +51,13 @@ public class PeersInformation {
     private final PeerScoringManager peerScoringManager;
     private final Comparator<Map.Entry<Peer, SyncPeerStatus>> peerComparator;
     private Map<Peer, SyncPeerStatus> peerStatuses = new HashMap<>();
+    private final double percentageOfPeersToConsiderInRandomSelection;
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:54]: it seems that this config could be part of `SyncConfiguration` class? Also can it be more shorter name? :) Eg. `topBest`?

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java [hunk 4]
```diff
@@ -106,6 +109,25 @@ public SyncPeerStatus getPeer(Peer peer) {
     }
 
     public Optional<Peer> getBestPeer() {
+        if (percentageOfPeersToConsiderInRandomSelection > 0.0D) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:112]: could you please cover edge cases of this part with unit tests?

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java [hunk 5]
```diff
@@ -106,6 +109,25 @@ public SyncPeerStatus getPeer(Peer peer) {
...
+            List<Map.Entry<Peer, SyncPeerStatus>> entries = getBestCandidatesStream()
+                    .sorted(this.peerComparator.reversed())
+                    .collect(Collectors.toList());
+
+            if (entries.isEmpty()) {
+                return Optional.empty();
+            }
+
+            int numberOfPeersToConsider = (int) Math.floor(entries.size() * (percentageOfPeersToConsiderInRandomSelection / 100));
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:121]: it seems that `numberOfPeersToConsider ` would equal to zero even if we have one peer and, for instance, percentage equals to 40%?

**Review by Vovchyk — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/java/org/ethereum/config/SystemProperties.java (hunk 1 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:101]: No reason to be `rpc` I don't know the whole config structure. I can move it to be under `sync`

Comment (rmoreliovlabs) [reply rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:101]: Addressed

File: rskj-core/src/main/java/org/ethereum/config/SystemProperties.java (hunk 2 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:722]: Renamed

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java (hunk 3 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:54]: Done

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java (hunk 5 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:121]: Addressed

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java (hunk 4 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:112]: I will add unit tests when we agree on the implementation

File: rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java [hunk 6]
```diff
@@ -24,10 +24,10 @@
 
 @Immutable
 public final class SyncConfiguration {
-    public static final SyncConfiguration DEFAULT = new SyncConfiguration(5, 60, 30, 5, 20, 192, 20, 10);
+    public static final SyncConfiguration DEFAULT = new SyncConfiguration(5, 60, 30, 5, 20, 192, 20, 10, 0);
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java:27]: seems it's only used in tests. `@VisibleForTesting` ?

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java [hunk 7]
```diff
@@ -51,11 +51,15 @@ public class PeersInformation {
     private final PeerScoringManager peerScoringManager;
     private final Comparator<Map.Entry<Peer, SyncPeerStatus>> peerComparator;
     private Map<Peer, SyncPeerStatus> peerStatuses = new HashMap<>();
+    private final double topBest;
+    private final Random random = new Random();
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:55]: `SecureRandom` ?

File: rskj-core/src/main/java/co/rsk/RskContext.java [hunk 8]
```diff
@@ -1886,11 +1887,14 @@ protected SyncProcessor getSyncProcessor() {
...
+            SyncConfiguration syncConfiguration = getSyncConfiguration();
+
             peersInformation = new PeersInformation(
                     getChannelManager(),
-                    getSyncConfiguration(),
+                    syncConfiguration,
                     getBlockchain(),
-                    getPeerScoringManager());
+                    getPeerScoringManager(),
+                    syncConfiguration.getTopBest());
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/RskContext.java:1897]: we alrady pass `syncConfiguration` as a param to the `PeersInformation` ctor. This extra param might not be needed

File: rskj-core/src/main/resources/reference.conf [hunk 9]
```diff
@@ -306,6 +306,9 @@ sync {
     # Amount of blocks contained in a chunk,
     # MUST BE 192 or a divisor of 192
     chunkSize = 192
+
+    # X % of top best nodes will be considered when for random selection
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/reference.conf:310]:
```suggestion
    # X % of top best nodes will be considered for random selection
```

File: rskj-core/src/main/resources/reference.conf [hunk 10]
```diff
@@ -306,6 +306,9 @@ sync {
...
+    topBest = 40
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/reference.conf:311]: perhaps it should be zero by default

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java [hunk 11]
```diff
@@ -106,6 +111,24 @@ public SyncPeerStatus getPeer(Peer peer) {
...
+        if (topBest > 0.0D) {
+            List<Map.Entry<Peer, SyncPeerStatus>> entries = getBestCandidatesStream()
+                    .sorted(this.peerComparator.reversed())
+                    .collect(Collectors.toList());
+
+            if (entries.isEmpty()) {
+                return Optional.empty();
+            }
+
+            int numberOfPeersToConsider = (int) Math.round(((double) entries.size()) * (topBest / 100.0D));
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:123]: I'd double check edge cases here, eg. when this results to `zero` or to `entries.size()`. You may wanna use `Math.max(1, <...>)` and `Math.min(entries.size(), <...>)`. You may also want to validate `topBest` config value somewhere that it's in the `[0..100]` range - probably in `RskSystemProperties`

**Review by Vovchyk — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java (hunk 6 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java:27]: Addressed

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java (hunk 7 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:55]: Addressed

File: rskj-core/src/main/java/co/rsk/RskContext.java (hunk 8 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/RskContext.java:1897]: Addressed

File: rskj-core/src/main/resources/reference.conf (hunk 10 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/resources/reference.conf:311]: Addressded

File: rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java (hunk 11 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/net/sync/PeersInformation.java:123]: Addressed

File: rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java [hunk 12]
```diff
@@ -47,6 +49,7 @@ public final class SyncConfiguration {
      * @param chunkSize Amount of blocks contained in a chunk
      * @param maxRequestedBodies Amount of bodies to request at the same time when synchronizing backwards.
      * @param longSyncLimit Distance to the tip of the peer's blockchain to enable long synchronization.
+     * @param topBest Distance to the tip of the peer's blockchain to enable long synchronization.
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java:52]: copy-paste issue?

File: rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java [hunk 13]
```diff
@@ -418,4 +420,18 @@ public Integer getMessageQueueMaxSize() {
...
+
+    public double getTopBest() {
+        if (!configFromFiles.hasPath(PROPERTY_SYNC_TOP_BEST)) {
+            return 0.0D;
+        }
+
+        double value = configFromFiles.getDouble(PROPERTY_SYNC_TOP_BEST);
+
+        if (value >= 0.0D && value <= 100.0D) {
+            throw new RuntimeException(PROPERTY_SYNC_TOP_BEST + " must be between 0 and 100");
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java:432]: `RskConfigurationException` ?

File: rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java [hunk 14]
```diff
@@ -418,4 +420,18 @@ public Integer getMessageQueueMaxSize() {
...
     }
+
+    public double getTopBest() {
+        if (!configFromFiles.hasPath(PROPERTY_SYNC_TOP_BEST)) {
+            return 0.0D;
+        }
+
+        double value = configFromFiles.getDouble(PROPERTY_SYNC_TOP_BEST);
+
+        if (value >= 0.0D && value <= 100.0D) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java:431]: probably 0 and 100 are legit values, wdyt?

File: rskj-core/src/main/resources/reference.conf (hunk 10 above)
Comment (Vovchyk) [reply rskj-core/src/main/resources/reference.conf:311]: it's still not zero

File: rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java (hunk 12 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/net/sync/SyncConfiguration.java:52]: Addressed

File: rskj-core/src/main/resources/reference.conf (hunk 10 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/resources/reference.conf:311]: Now should be

File: rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java (hunk 13 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java:432]: Addressed

File: rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java (hunk 14 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java:431]: Addressed

**Review by Vovchyk — DISMISSED** [review]

**Comment by Vovchyk** [comment]
pipeline:run

**Review by Vovchyk — APPROVED** [review]
